src/openloco/windows/constructionwnd.cpp

```diff
@@ -1,75 +1,4836 @@
+        auto stationListCopy = _stationList;
+        auto currentYear = current_year();
+        auto trackObj = objectmgr::get<track_object>(trackType);
+        auto trainStationCount = 0;
+
+        for (auto i = 0; i < trackObj->num_stations; i++)
+        {
+            auto station = trackObj->stations[i];
+            if (station == 0xFF)
+                continue;
+            auto trainStationObj = objectmgr::get<train_station_object>(station);
+            if (currentYear < trainStationObj->designed_year)
+                continue;
+            if (currentYear > trainStationObj->obsolete_year)
+                continue;
+            _stationList[trainStationCount] = station;
+            trainStationCount++;
+        }
+
+        for (uint8_t i = 0; i < objectmgr::get_max_objects(object_type::track_station); i++)
+        {
+            auto trainStationObj = objectmgr::get<train_station_object>(i);
+            if (trainStationObj == nullptr)
+                continue;
+            for (auto j = 0; j < trainStationObj->var_22; j++)
+            {
+                if (trackType != trainStationObj->num_compatible)
+                    continue;
+                if (currentYear < trainStationObj->designed_year)
+                    continue;
+                if (currentYear > trainStationObj->obsolete_year)
+                    continue;
+                for (size_t k = 0; k < std::size(_stationList); k++)
+                {
+                    auto station = stationListCopy[k];
+                    if (stationListCopy == _stationList)
+                    {
+                        _stationList[trainStationCount] = i;
+                        trainStationCount++;
+                        break;
+                    }
+                    if (i == station)
+                        break;
+                }
+            }
+        }
+
+        _stationList[trainStationCount] = 0xFF;
+
+        sortList(_stationList, std::size(_stationList));
+    }
+
+    // 0x0042C490
+    static void refreshTrackBridgeList()
+    {
+        auto trackType = _trackType;
+        auto bridgeListCopy = _bridgeList;
+        auto currentYear = current_year();
+        auto trackObj = objectmgr::get<track_object>(trackType);
+        auto bridgeCount = 0;
+        for (auto i = 0; i < trackObj->num_bridges; i++)
+        {
+            auto bridge = trackObj->bridges[i];
+            if (bridge == 0xFF)
+                continue;
+            auto bridgeObj = objectmgr::get<bridge_object>(bridge);
+            if (currentYear < bridgeObj->designed_year)
+                continue;
+            _bridgeList[bridgeCount] = bridge;
+            bridgeCount++;
+        }
+
+        for (uint8_t i = 0; i < objectmgr::get_max_objects(object_type::bridge); i++)
+        {
+            auto bridgeObj = objectmgr::get<bridge_object>(i);
+            if (bridgeObj == nullptr)
+                continue;
+            for (auto j = 0; j < bridgeObj->var_22; j++)
```

Comment:
This is incorrect the track has a different num_compatible to roads its at var_1A and the array at var_1B